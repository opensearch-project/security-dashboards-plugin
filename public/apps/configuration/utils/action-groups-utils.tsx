/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { HttpStart } from 'opensearch-dashboards/public';
import { map } from 'lodash';
import { API_ENDPOINT_ACTIONGROUPS, CLUSTER_PERMISSIONS, INDEX_PERMISSIONS } from '../constants';
import { DataObject, ActionGroupItem, ActionGroupUpdate, ObjectsMessage } from '../types';
import { httpDelete, httpGet, httpPost } from './request-utils';
import { getResourceUrl } from './resource-utils';

export interface PermissionListingItem {
  name: string;
  type: 'Action group' | 'Single permission';
  reserved: boolean;
  allowedActions: string[];
  hasClusterPermission: boolean;
  hasIndexPermission: boolean;
}

export async function fetchActionGroups(http: HttpStart): Promise<DataObject<ActionGroupItem>> {
  const actiongroups = await httpGet<ObjectsMessage<ActionGroupItem>>({
    http,
    url: API_ENDPOINT_ACTIONGROUPS,
  });
  return actiongroups.data;
}

export function transformActionGroupsToListingFormat(
  rawData: DataObject<ActionGroupItem>
): PermissionListingItem[] {
  return map(rawData, (value: ActionGroupItem, key?: string) => ({
    name: key || '',
    type: 'Action group',
    reserved: value.reserved,
    allowedActions: value.allowed_actions,
    hasClusterPermission: value.type === 'cluster' || value.type === 'all',
    hasIndexPermission: value.type === 'index' || value.type === 'all',
  }));
}

export async function fetchActionGroupListing(http: HttpStart): Promise<PermissionListingItem[]> {
  return transformActionGroupsToListingFormat(await fetchActionGroups(http));
}

function getClusterSinglePermissions(): PermissionListingItem[] {
  return CLUSTER_PERMISSIONS.map((permission) => ({
    name: permission,
    type: 'Single permission',
    reserved: true,
    allowedActions: [],
    hasClusterPermission: true,
    hasIndexPermission: false,
  }));
}

function getIndexSinglePermissions(): PermissionListingItem[] {
  return INDEX_PERMISSIONS.map((permission) => ({
    name: permission,
    type: 'Single permission',
    reserved: true,
    allowedActions: [],
    hasClusterPermission: false,
    hasIndexPermission: true,
  }));
}

export async function getAllPermissionsListing(http: HttpStart): Promise<PermissionListingItem[]> {
  return mergeAllPermissions(await fetchActionGroups(http));
}

export async function mergeAllPermissions(
  actionGroups: DataObject<ActionGroupItem>
): Promise<PermissionListingItem[]> {
  return transformActionGroupsToListingFormat(actionGroups)
    .concat(getClusterSinglePermissions())
    .concat(getIndexSinglePermissions());
}

export async function updateActionGroup(
  http: HttpStart,
  groupName: string,
  updateObject: ActionGroupUpdate
): Promise<ActionGroupUpdate> {
  return await httpPost({
    http,
    url: getResourceUrl(API_ENDPOINT_ACTIONGROUPS, groupName),
    body: updateObject,
  });
}

export async function requestDeleteActionGroups(http: HttpStart, groups: string[]) {
  for (const group of groups) {
    await httpDelete({ http, url: getResourceUrl(API_ENDPOINT_ACTIONGROUPS, group) });
  }
}
