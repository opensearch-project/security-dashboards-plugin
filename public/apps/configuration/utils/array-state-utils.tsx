/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { get, set, curry, PropertyName } from 'lodash';
import { Dispatch, SetStateAction } from 'react';

function resolveValue<T>(value: T | (() => T)) {
  return typeof value === 'function' ? (value as Function)() : value;
}

/**
 * Update an element in an sub array
 * @param setStateCallback setState function
 * @param path lodash path, e.g. [0, indexPattern] or "[0].indexPattern"
 * @param newValue value to be updated
 *
 * e.g.
 *  currentState = [{a: 1}, {a: 2}]
 *  path = [0, 'a']
 *  newValue = 10
 * The newState will be [{a: 10}, {a: 2}]
 */
export function updateElementInArray<T>(
  setStateCallback: Dispatch<SetStateAction<any[]>>,
  path: PropertyName | PropertyName[],
  newValue: T | (() => T)
) {
  setStateCallback((prevState) => {
    const newState = [...prevState];
    set(newState, path, resolveValue(newValue));
    return newState;
  });
}

export const updateElementInArrayHandler = curry(updateElementInArray);

/**
 * Append an element to an array (or sub array indicating by path)
 * @param setStateCallback setState function
 * @param path lodash path, e.g. [0, indexPattern] or "[0].indexPattern", use [] to indicate root level
 * @param newValue value to be updated
 *
 * e.g.
 * Scenario 1, path = [] to append to root level array
 *  currentState = [1, 2]
 *  path = []
 *  newValue = 3
 * The newState will be [1, 2, 3]
 *
 * Scenario 2, path != [] to append to sub array
 *  currentState = [{a: [1, 2]}, {a: [3, 4]}]
 *  path = [0, 'a']
 *  newValue = 5
 * The newState will be [{a: [1, 2, 5]}, {a: [3, 4]}]
 */
export function appendElementToArray<T>(
  setStateCallback: Dispatch<SetStateAction<any[]>>,
  path: PropertyName | PropertyName[],
  newValue: T | (() => T)
) {
  const resolvedNewValue = resolveValue(newValue);
  setStateCallback((prevState) => {
    if ((path as PropertyName[]).length === 0) {
      return [...prevState, resolvedNewValue];
    } else {
      const newArray = [...(get(prevState, path) as T[]), resolvedNewValue];
      const newState = [...prevState];
      set(newState, path, newArray);
      return newState;
    }
  });
}

/**
 * Remove an element from an array (or sub array indicating by path)
 * @param setStateCallback setState function
 * @param path lodash path, e.g. [0, indexPattern] or "[0].indexPattern", use [] to indicate root level
 * @param index index of element to be removed
 *
 * e.g.
 * Scenario 1, path = [] to append to root level array
 *  currentState = [1, 2]
 *  path = []
 *  index = 0
 * The newState will be [2]
 *
 * Scenario 2, path != [] to append to sub array
 *  currentState = [{a: [1, 2]}, {a: [3, 4]}]
 *  path = [0, 'a']
 *  index = 1
 * The newState will be [{a: [1]}, {a: [3, 4]}]
 */
export function removeElementFromArray<T>(
  setStateCallback: Dispatch<SetStateAction<any[]>>,
  path: PropertyName | PropertyName[],
  index: number
) {
  setStateCallback((prevState) => {
    if ((path as PropertyName[]).length === 0) {
      const newState = [...prevState];
      newState.splice(index, 1);
      return newState;
    } else {
      const newArray = [...(get(prevState, path) as T[])];
      newArray.splice(index, 1);
      const newState = [...prevState];
      set(newState, path, newArray);
      return newState;
    }
  });
}
