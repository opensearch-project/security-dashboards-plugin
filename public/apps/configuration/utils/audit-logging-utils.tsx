/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { HttpStart } from 'opensearch-dashboards/public';
import { AuditLoggingSettings } from '../panels/audit-logging/types';
import { API_ENDPOINT_AUDITLOGGING, API_ENDPOINT_AUDITLOGGING_UPDATE } from '../constants';
import { httpGet, httpPost } from './request-utils';

export async function updateAuditLogging(http: HttpStart, updateObject: AuditLoggingSettings) {
  return await httpPost({ http, url: API_ENDPOINT_AUDITLOGGING_UPDATE, body: updateObject });
}

export async function getAuditLogging(http: HttpStart): Promise<AuditLoggingSettings> {
  const rawConfiguration = await httpGet<any>({ http, url: API_ENDPOINT_AUDITLOGGING });
  return rawConfiguration?.config;
}
