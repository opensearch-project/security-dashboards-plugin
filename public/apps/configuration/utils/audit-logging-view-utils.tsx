/*
 *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */
import {
  EuiDescribedFormGroup,
  EuiFlexGrid,
  EuiForm,
  EuiFormRow,
  EuiHorizontalRule,
  EuiPanel,
  EuiSpacer,
  EuiSwitch,
  EuiText,
  EuiTitle,
} from '@elastic/eui';

import React from 'react';
import { HttpStart } from 'kibana/public';
import { displayArray, displayBoolean, displayObject, renderTextFlexItem } from './display-utils';
import { API_ENDPOINT_AUDITLOGGING_UPDATE } from '../constants';
import {
  AuditLoggingSettings,
  ComplianceSettings,
  GeneralSettings,
} from '../panels/audit-logging/types';

export function renderStatusPanel(onSwitchChange: () => void, auditLoggingEnabled: boolean) {
  return (
    <EuiPanel>
      <EuiTitle>
        <h3>Audit logging</h3>
      </EuiTitle>
      <EuiHorizontalRule />
      <EuiForm>
        <EuiDescribedFormGroup
          title={<h3>Enable audit logging</h3>}
          description={<>Enable or disable audit logging</>}
        >
          <EuiFormRow>
            <EuiSwitch
              name="auditLoggingEnabledSwitch"
              label={displayBoolean(auditLoggingEnabled)}
              checked={auditLoggingEnabled}
              onChange={onSwitchChange}
            />
          </EuiFormRow>
        </EuiDescribedFormGroup>
      </EuiForm>
    </EuiPanel>
  );
}

export function renderGeneralSettings(generalSettings: GeneralSettings) {
  return (
    <>
      {renderGeneralSettingsForLayer(generalSettings)}

      <EuiSpacer />

      {renderGeneralSettingsForAttribute(generalSettings)}

      <EuiSpacer />

      {renderGeneralSettingsForIgnore(generalSettings)}
    </>
  );
}

function renderGeneralSettingsForLayer(generalSettings: GeneralSettings) {
  return (
    <>
      <EuiText>
        <h3>Layer settings</h3>
      </EuiText>

      <EuiSpacer />

      <EuiFlexGrid columns={3}>
        {renderTextFlexItem('REST layer', displayBoolean(generalSettings.enable_rest))}
        {renderTextFlexItem('Transport layer', displayBoolean(generalSettings.enable_transport))}
        {renderTextFlexItem(
          'REST disabled categories',
          displayArray(generalSettings.disabled_rest_categories)
        )}
        {renderTextFlexItem(
          'Transport disabled categories',
          displayArray(generalSettings.disabled_transport_categories)
        )}
      </EuiFlexGrid>
    </>
  );
}

function renderGeneralSettingsForAttribute(generalSettings: GeneralSettings) {
  return (
    <>
      <EuiText>
        <h3>Attribute settings</h3>
      </EuiText>

      <EuiSpacer />

      <EuiFlexGrid columns={3}>
        {renderTextFlexItem('Bulk requests', displayBoolean(generalSettings.resolve_bulk_requests))}
        {renderTextFlexItem('Resolve indices', displayBoolean(generalSettings.resolve_indices))}
        {renderTextFlexItem('Request body', displayBoolean(generalSettings.log_request_body))}
        {renderTextFlexItem(
          'Sensitive headers',
          displayBoolean(generalSettings.exclude_sensitive_headers)
        )}
      </EuiFlexGrid>
    </>
  );
}

function renderGeneralSettingsForIgnore(generalSettings: GeneralSettings) {
  return (
    <>
      <EuiText>
        <h3>Ignore settings</h3>
      </EuiText>

      <EuiSpacer />

      <EuiFlexGrid columns={3}>
        {renderTextFlexItem('Exclude users', displayArray(generalSettings.ignore_users))}
        {renderTextFlexItem('Exclude requests', displayArray(generalSettings.ignore_requests))}
      </EuiFlexGrid>
    </>
  );
}

export function renderComplianceSettings(complianceSettings: ComplianceSettings) {
  return (
    <>
      {renderTextFlexItem('Compliance mode', displayBoolean(complianceSettings.enabled))}

      <EuiSpacer />

      <EuiPanel>
        <EuiTitle>
          <h4>Read</h4>
        </EuiTitle>

        <EuiSpacer />

        <EuiFlexGrid columns={3}>
          {renderTextFlexItem(
            'Read metadata only',
            displayBoolean(complianceSettings.read_metadata_only)
          )}
          {renderTextFlexItem('Ignored users', displayArray(complianceSettings.read_ignore_users))}
          {/* read_watched_fields is object instead of array.*/}
          {renderTextFlexItem(
            'Watched fields',
            displayObject(complianceSettings.read_watched_fields)
          )}
        </EuiFlexGrid>
      </EuiPanel>

      <EuiSpacer />

      <EuiPanel>
        <EuiTitle>
          <h4>Write</h4>
        </EuiTitle>

        <EuiSpacer />

        <EuiFlexGrid columns={3}>
          {renderTextFlexItem(
            'Write metadata only',
            displayBoolean(complianceSettings.write_metadata_only)
          )}
          {renderTextFlexItem('Log diffs', displayBoolean(complianceSettings.write_log_diffs))}
          {renderTextFlexItem('Ignored users', displayArray(complianceSettings.write_ignore_users))}
          {renderTextFlexItem(
            'Watched indices',
            displayArray(complianceSettings.write_watched_indices)
          )}
        </EuiFlexGrid>
      </EuiPanel>
    </>
  );
}

export async function updateAuditLogging(http: HttpStart, updateObject: AuditLoggingSettings) {
  return await http.post(`${API_ENDPOINT_AUDITLOGGING_UPDATE}`, {
    body: JSON.stringify(updateObject),
  });
}
