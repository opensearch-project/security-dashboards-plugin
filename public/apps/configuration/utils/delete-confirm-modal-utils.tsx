/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React, { useState } from 'react';
import { EuiOverlayMask, EuiConfirmModal } from '@elastic/eui';

/**
 *
 * @param handleDelete: [Type: func] delete function which needs to be execute on click of confirm button.
 * @param entity: e.g. role(s), tenant(s), user(s), mapping etc. This will display in confirmation text before deletion.
 * @param customConfirmationText: If you want other than default confirm message, pass it as customConfirmationText.
 */
export function useDeleteConfirmState(
  handleDelete: () => Promise<void>,
  entity: string,
  customConfirmationText?: React.ReactNode
): [() => void, React.ReactNode] {
  const [isDeleteConfirmModalVisible, setIsDeleteConfirmModalVisible] = useState(false);
  const closeDeleteConfirmModal = () => setIsDeleteConfirmModalVisible(false);
  const showDeleteConfirmModal = () => setIsDeleteConfirmModalVisible(true);

  const handleConfirm = async () => {
    await handleDelete();
    closeDeleteConfirmModal();
  };

  let deleteConfirmModal;
  if (isDeleteConfirmModalVisible) {
    deleteConfirmModal = (
      <EuiOverlayMask>
        <EuiConfirmModal
          title="Confirm Delete"
          onCancel={closeDeleteConfirmModal}
          onConfirm={handleConfirm}
          cancelButtonText="Cancel"
          confirmButtonText="Confirm"
          defaultFocusedButton="confirm"
        >
          {customConfirmationText ? (
            customConfirmationText
          ) : (
            <p>Do you really want to delete selected {entity}?</p>
          )}
        </EuiConfirmModal>
      </EuiOverlayMask>
    );
  }
  return [showDeleteConfirmModal, deleteConfirmModal];
}
