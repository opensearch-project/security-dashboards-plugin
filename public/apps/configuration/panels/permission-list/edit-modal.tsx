/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import {
  EuiButton,
  EuiButtonEmpty,
  EuiModal,
  EuiModalBody,
  EuiModalFooter,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiOverlayMask,
  EuiForm,
  EuiComboBox,
} from '@elastic/eui';
import React, { useState } from 'react';
import { ComboBoxOptions, Action } from '../../types';
import {
  stringToComboBoxOption,
  comboBoxOptionToString,
  appendOptionToComboBoxHandler,
} from '../../utils/combo-box-utils';
import { FormRow } from '../../utils/form-row';
import { NameRow } from '../../utils/name-row';

interface PermissionEditModalDeps {
  groupName: string;
  action: Action;
  allowedActions: string[];
  optionUniverse: ComboBoxOptions;
  handleClose: () => void;
  handleSave: (groupName: string, allowedActions: string[]) => Promise<void>;
}

const TITLE_DICT: { [key: string]: string } = {
  [Action.create]: 'Create new action group',
  [Action.edit]: 'Edit action group',
  [Action.duplicate]: 'Duplicate action group',
};

export function PermissionEditModal(props: PermissionEditModalDeps) {
  const [groupName, setGroupName] = useState<string>(props.groupName);
  const [allowedActions, setAllowedActions] = useState<ComboBoxOptions>(
    props.allowedActions.map(stringToComboBoxOption)
  );
  const [isFormValid, setIsFormValid] = useState<boolean>(true);

  return (
    <EuiOverlayMask>
      <EuiModal onClose={props.handleClose}>
        <EuiModalHeader>
          <EuiModalHeaderTitle>{TITLE_DICT[props.action]}</EuiModalHeaderTitle>
        </EuiModalHeader>

        <EuiModalBody>
          <EuiForm>
            <NameRow
              headerText="Name"
              headerSubText="Enter a unique name to describe the purpose of this group. You cannot change the name after creation."
              resourceName={groupName}
              resourceType=""
              action={props.action}
              setNameState={setGroupName}
              setIsFormValid={setIsFormValid}
            />
            <FormRow headerText="Permissions">
              <EuiComboBox
                options={props.optionUniverse}
                selectedOptions={allowedActions}
                onCreateOption={appendOptionToComboBoxHandler(setAllowedActions, [])}
                onChange={setAllowedActions}
              />
            </FormRow>
          </EuiForm>
        </EuiModalBody>

        <EuiModalFooter>
          <EuiButtonEmpty onClick={props.handleClose}>Cancel</EuiButtonEmpty>

          <EuiButton
            id="submit"
            onClick={async () => {
              await props.handleSave(groupName, allowedActions.map(comboBoxOptionToString));
            }}
            fill
            disabled={!isFormValid}
          >
            {props.action === Action.create ? 'Create' : 'Save'}
          </EuiButton>
        </EuiModalFooter>
      </EuiModal>
    </EuiOverlayMask>
  );
}
