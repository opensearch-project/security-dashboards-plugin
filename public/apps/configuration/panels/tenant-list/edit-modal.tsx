/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import {
  EuiButton,
  EuiButtonEmpty,
  EuiModal,
  EuiModalBody,
  EuiModalFooter,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiOverlayMask,
  EuiForm,
  EuiTextArea,
  EuiHorizontalRule,
} from '@elastic/eui';
import React, { useState } from 'react';
import { Action } from '../../types';
import { FormRow } from '../../utils/form-row';
import { NameRow } from '../../utils/name-row';

interface TenantEditModalDeps {
  tenantName: string;
  tenantDescription: string;
  action: Action;
  handleClose: () => void;
  handleSave: (tenantName: string, tenantDescription: string) => Promise<void>;
}

const TITLE_DICT: { [key: string]: string } = {
  [Action.create]: 'Create tenant',
  [Action.edit]: 'Edit tenant',
  [Action.duplicate]: 'Duplicate tenant',
};

export function TenantEditModal(props: TenantEditModalDeps) {
  const [tenantName, setTenantName] = useState<string>(props.tenantName);
  const [tenantDescription, setTenantDescription] = React.useState<string>(props.tenantDescription);

  const [isFormValid, setIsFormValid] = useState<boolean>(true);

  const handleTenantDescriptionChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    setTenantDescription(e.target.value);
  };

  return (
    <EuiOverlayMask>
      <EuiModal onClose={props.handleClose}>
        <EuiModalHeader>
          <EuiModalHeaderTitle>{TITLE_DICT[props.action]}</EuiModalHeaderTitle>
        </EuiModalHeader>

        <EuiModalBody>
          <EuiForm>
            <NameRow
              headerText="Name"
              headerSubText="Specify a descriptive and unique tenant name that is easy to recognize. You cannot edit the name once the tenant is created."
              resourceName={tenantName}
              resourceType="tenant"
              action={props.action}
              setNameState={setTenantName}
              setIsFormValid={setIsFormValid}
              fullWidth={true}
            />
            <FormRow
              headerText="Description"
              headerSubText="Describe the purpose of the tenant."
              optional
            >
              <EuiTextArea
                data-test-subj="tenant-description"
                fullWidth
                placeholder="Describe the tenant"
                value={tenantDescription}
                onChange={handleTenantDescriptionChange}
              />
            </FormRow>
          </EuiForm>
        </EuiModalBody>
        <EuiHorizontalRule margin="xs" />
        <EuiModalFooter>
          <EuiButtonEmpty onClick={props.handleClose}>Cancel</EuiButtonEmpty>

          <EuiButton
            id="submit"
            onClick={async () => {
              await props.handleSave(tenantName, tenantDescription);
            }}
            fill
            disabled={!isFormValid}
          >
            {props.action === Action.create ? 'Create' : 'Save'}
          </EuiButton>
        </EuiModalFooter>
      </EuiModal>
    </EuiOverlayMask>
  );
}
