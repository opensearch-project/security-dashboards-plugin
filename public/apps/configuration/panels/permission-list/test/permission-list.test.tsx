/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { shallow } from 'enzyme';
import React from 'react';
import {
  PermissionList,
  renderBooleanToCheckMark,
  toggleRowDetails,
  renderRowExpansionArrow,
} from '../permission-list';
import { EuiInMemoryTable, EuiButtonIcon } from '@elastic/eui';
import {
  requestDeleteActionGroups,
  fetchActionGroups,
  updateActionGroup,
  PermissionListingItem,
} from '../../../utils/action-groups-utils';
import { useDeleteConfirmState } from '../../../utils/delete-confirm-modal-utils';
import { PermissionEditModal } from '../edit-modal';

jest.mock('../../../utils/action-groups-utils', () => ({
  requestDeleteActionGroups: jest.fn(),
  fetchActionGroups: jest.fn(),
  mergeAllPermissions: jest.fn().mockReturnValue([]),
  updateActionGroup: jest.fn(),
}));
jest.mock('../../../utils/delete-confirm-modal-utils', () => ({
  useDeleteConfirmState: jest.fn().mockReturnValue([jest.fn(), '']),
}));
jest.mock('../../../utils/context-menu', () => ({
  useContextMenuState: jest
    .fn()
    .mockImplementation((buttonText, buttonProps, children) => [children, jest.fn()]),
}));
jest.mock('../../../utils/toast-utils', () => ({
  useToastState: jest.fn().mockReturnValue([[], jest.fn(), jest.fn()]),
}));

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useContext: jest.fn().mockReturnValue({ dataSource: { id: 'test' }, setDataSource: jest.fn() }), // Mock the useContext hook to return dummy datasource and setdatasource function
}));

describe('Permission list page ', () => {
  const sampleActionGroup: PermissionListingItem = {
    name: 'group',
    type: 'Action group',
    reserved: false,
    allowedActions: [],
    hasClusterPermission: true,
    hasIndexPermission: false,
  };

  it('renderBooleanToCheckMark', () => {
    expect(renderBooleanToCheckMark(false)).toBeFalsy();
    expect(renderBooleanToCheckMark(true)).toBeTruthy();
  });

  it('toggleRowDetails', () => {
    const setMap = jest.fn();
    toggleRowDetails(sampleActionGroup, {}, setMap);
    const updateMapFunc = setMap.mock.calls[0][0];
    updateMapFunc({ group: '' });
  });

  describe('renderRowExpanstionArrow', () => {
    it('should render down arrow when collapsed', () => {
      const renderFunc = renderRowExpansionArrow({}, {}, jest.fn());
      const Wrapper = () => <>{renderFunc(sampleActionGroup)}</>;
      const component = shallow(<Wrapper />);

      expect(component.find(EuiButtonIcon).prop('iconType')).toBe('arrowDown');
    });

    it('should render up arrow when expanded', () => {
      const renderFunc = renderRowExpansionArrow(
        { [sampleActionGroup.name]: sampleActionGroup },
        {},
        jest.fn()
      );
      const Wrapper = () => <>{renderFunc(sampleActionGroup)}</>;
      const component = shallow(<Wrapper />);

      expect(component.find(EuiButtonIcon).prop('iconType')).toBe('arrowUp');
    });
  });

  describe('PermissionList', () => {
    const mockCoreStart = {
      http: 1,
    };
    it('render empty', () => {
      const component = shallow(
        <PermissionList
          coreStart={{} as any}
          depsStart={{} as any}
          params={{} as any}
          config={{} as any}
        />
      );

      expect(component.find(EuiInMemoryTable).prop('items').length).toBe(0);
    });

    it('fetch data', () => {
      jest.spyOn(React, 'useEffect').mockImplementationOnce((f) => f());
      shallow(
        <PermissionList
          coreStart={mockCoreStart as any}
          depsStart={{} as any}
          params={{} as any}
          config={{} as any}
        />
      );

      expect(fetchActionGroups).toBeCalledWith(mockCoreStart.http, 'test');
    });

    it('fetch data error', () => {
      const consoleLog = jest.spyOn(console, 'log');
      consoleLog.mockImplementation(() => {});
      jest.spyOn(React, 'useEffect').mockImplementationOnce((f) => f());
      const error = new Error();
      fetchActionGroups.mockImplementationOnce(() => {
        throw error;
      });
      // Hide the error message
      jest.spyOn(console, 'log').mockImplementationOnce(() => {});
      shallow(
        <PermissionList
          coreStart={{} as any}
          depsStart={{} as any}
          params={{} as any}
          config={{} as any}
        />
      );

      // Expect error log
      expect(consoleLog).toBeCalledWith(error);
    });

    it('submit change', () => {
      const component = shallow(
        <PermissionList
          coreStart={mockCoreStart as any}
          depsStart={{} as any}
          params={{} as any}
          config={{} as any}
        />
      );
      component.find('#create-from-selection').simulate('click');
      const submitFunc = component.find(PermissionEditModal).prop('handleSave');
      submitFunc('group1', []);

      expect(updateActionGroup).toBeCalledWith(
        mockCoreStart.http,
        'group1',
        { allowed_actions: [] },
        'test'
      );
    });

    it('submit change error', () => {
      const consoleLog = jest.spyOn(console, 'log');
      consoleLog.mockImplementation(() => {});
      const error = new Error();
      updateActionGroup.mockImplementationOnce(() => {
        throw error;
      });
      const component = shallow(
        <PermissionList
          coreStart={{} as any}
          depsStart={{} as any}
          params={{} as any}
          config={{} as any}
        />
      );
      component.find('#create-from-selection').simulate('click');
      const submitFunc = component.find(PermissionEditModal).prop('handleSave');
      submitFunc('group1', []);

      // Expect error log
      expect(consoleLog).toBeCalledWith(error);
    });

    it('delete action group', (done) => {
      shallow(
        <PermissionList
          coreStart={mockCoreStart as any}
          depsStart={{} as any}
          params={{} as any}
          config={{} as any}
        />
      );
      const deleteFunc = useDeleteConfirmState.mock.calls[0][0];

      deleteFunc();

      process.nextTick(() => {
        expect(requestDeleteActionGroups).toBeCalledWith(mockCoreStart.http, [], 'test');
        done();
      });
    });

    it('edit and duplicate button should be enabled when there is 1 custom action group selected', () => {
      jest.spyOn(React, 'useState').mockImplementation(() => [[sampleActionGroup], jest.fn()]);
      const component = shallow(
        <PermissionList
          coreStart={{} as any}
          depsStart={{} as any}
          params={{} as any}
          config={{} as any}
        />
      );

      expect(component.find('#edit').prop('disabled')).toBe(false);
      expect(component.find('#duplicate').prop('disabled')).toBe(false);
    });
  });

  describe('AccessError component', () => {
    const mockCoreStart = {
      http: 1,
    };
    let component;
    beforeEach(() => {
      fetchActionGroups.mockRejectedValueOnce({ response: { status: 403 } });
      jest.spyOn(React, 'useEffect').mockImplementationOnce((f) => f());
      component = shallow(
        <PermissionList
          coreStart={mockCoreStart as any}
          depsStart={{} as any}
          params={{} as any}
          config={{} as any}
        />
      );
    });

    it('should load access error component', () => {
      expect(component).toMatchSnapshot();
    });
  });
});
