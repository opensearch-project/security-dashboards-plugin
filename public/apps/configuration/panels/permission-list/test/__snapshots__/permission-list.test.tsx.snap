// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Permission list page  AccessError component should load access error component 1`] = `
<Fragment>
  <Memo()
    config={Object {}}
    coreStart={
      Object {
        "http": 1,
      }
    }
    dataSourcePickerReadOnly={false}
    depsStart={Object {}}
    params={Object {}}
    selectedDataSource={
      Object {
        "id": "test",
      }
    }
    setDataSource={[MockFunction]}
  />
  <EuiPageHeader>
    <EuiTitle
      size="l"
    >
      <h1>
        Permissions
      </h1>
    </EuiTitle>
  </EuiPageHeader>
  <AccessErrorComponent
    loading={false}
  />
  <EuiGlobalToastList
    dismissToast={[MockFunction]}
    side="right"
    toastLifeTimeMs={10000}
    toasts={Array []}
  />
</Fragment>
`;
